### ISSUE TITLE
Update automated tests for CHT and OpenMRS endpoints in the mediator

### ISSUE THREAD

**Issue by witash** [medic/cht-interoperability#123]
Automated tests cover only FHIR endpoints.
Now that there are additional endpoints for OpenMRS and CHT, these need to be added to automated tests.

**Comment by lorerod** [comment]
@witash, I'm starting to work on this so you can focus on other needed developments. 
Let me know if this is ok with you.

**Comment by lorerod** [comment]
@witash I have a question about the `CHT Patient to OpenMRS diagram`. Is the entire flow triggered only by the creation of the patient in the CHT?

**Comment by witash** [comment]
Yes, it was, but #124 and #125 changed this; the point was to decoupe cht and openmrs, so cht only sends requests to the FHIR Server, and the same for OpenMRS; neither one knows or cares that the other side exists. So for patients, the only difference between this branch and the LTFU is that cht is sending raw documents instead of already FHIR formatted requests. And it creates Observation Resources as part of Encounters when CHT forms are submitted.

updated sequence diagrams:
[for patients created in cht](https://github.com/medic/cht-interoperability/blob/openmrs-mediator/docs/sequence-diagram/cht-patient-creation.png)
[for form submissions in cht](https://github.com/medic/cht-interoperability/blob/openmrs-mediator/docs/sequence-diagram/cht-form-submission.png)

Then, for OpenMRS, it doesn't have any subscription or event based messaging yet, so the mediator needs to poll it.
[here](https://github.com/medic/cht-interoperability/blob/openmrs-mediator/docs/sequence-diagram/cht-outgoing-patients.png) is the sequence diagram for [that](https://github.com/medic/cht-interoperability/blob/openmrs-mediator/mediator/src/utils/openmrs_sync.ts). Its also supposed to be a two-way sync, so in adddition to sending new resources from FHIR to OpenMRS, it gets new resources from OpenMRS to FHIR. theres no diagram for that yet.


So. for end to end tests we need
1. Patient Document sent from outbound push gets saved in FHIR Server
- [ document and add cht configs to configurator]
- [ postman collections for outbound push requests ]  
3. Encounter Document sent from outbound push gets saved in FHIR Server (w/ fields encoded as observations)
- [ document and add cht configs to configurator]
- [ postman collections for outbound push requests ]
4. New Patients in FHIR are sent to OpenMRS
5. New Patients in OpenMRS are sent to FHIR
6. New Encounters in FHIR are sent to OpenMRS - there's a different format and patient ids may be different
- [ document different format ]
- [ postman collections ] 
8. New Encounters in OpenMRS are sent FHIR
9. New Encounters in FHIR are sent to CHT - with observations attached and coded to cht form fields
 - [ document and add cht configs to configurator]
- [ postman collections for incoming form ]  
11. New Patients in FHIR are sent to CHT - with the right patient_ids and parent_id (place_id). 
- [ document and add cht configs to configurator]
- [ postman collections for outbound push requests ]

**Comment by lorerod** [comment]
Okay, thanks. Independent of the point of decoupling cht and openers, does it still make sense to implement e2e test scenarios representing the entire workflow?

**CHT to OpenMRS workflow**

1. Patient Document sent from outbound push gets saved in FHIR Server 
2. New Patients in FHIR are sent to OpenMRS 
3. Encounter Document sent from outbound push gets saved in FHIR Server (w/ fields encoded as observations)
4. New Encounters in FHIR are sent to OpenMRS - there's a different format and patient ids may be different

**OpenMRS Patient to CHT workflow**

1. New Patients in OpenMRS are sent to FHIR
2. New Patients in FHIR are sent to CHT - with the right patient_ids and parent_id (place_id).
3. New Encounters in OpenMRS are sent FHIR
4. New Encounters in FHIR are sent to CHT - with observations attached and coded to cht form fields

And also, does it make sense to combine both? New Patient in OpenMRS with New Encounter in CHT, testing the entire flow.

**Comment by lorerod** [comment]
Also, do your comments on "add cht configs to configurator" mean those are pending tasks you will implement? Or do I need to add the outbound push config as a precondition in the e2e tests?

**Comment by witash** [comment]
Also, do your comments on "add cht configs to configurator" mean those are pending tasks you will implement? Or do I need to add the outbound push config as a precondition in the e2e tests?
Those are pending tasks for me to implement. 

And also, does it make sense to combine both? for example: New Patient in OpenMRS with New Encounter in CHT, testing the entire flow.
Yes it easiest to combine them since encounters have dependencies on patients.

**Comment by witash** [comment]
endpoint and organization are not required; they don't have any functional significance. serviceRequests are not synced in this workflow; cht can create tasks from transitions after the patient form is submitted. if necessary, but they are not created by openmrs or any interoperability thing.

By `Form submission in CHT` I meant something like pregnancy registration; this is an addition to the LTFU workflow where if a CHW submits a pregnancy registration, or any other form thats [configured with outbound push](https://github.com/medic/config-gandaki/blob/b723bd12cee43aae15deefefba37f78eab99ab85/cht-config/app_settings/base_settings.json#L1916), it sends it to the mediator which converts it into an encounter, and the values to observations.

the equivalent of inter_follow_up form is a [form created in CHT config](https://github.com/medic/config-gandaki/blob/b723bd12cee43aae15deefefba37f78eab99ab85/cht-config/app_settings/forms.json#L1171), for incoming data; if someone on OpenMRS submits a form that is being tracked in CHT, the mediator goes the other way and converts the Observations and Encounter to the records api. Currently one of these has to be created for every corresponding form in OpenMRS, and also for [incoming patients](https://github.com/medic/config-gandaki/blob/b723bd12cee43aae15deefefba37f78eab99ab85/cht-config/app_settings/forms.json#L1064)

**Comment by lorerod** [comment]
@witash, this post is failing
`POST openhim-core:5001/mediator/openmrs/patient`
Do I need to add any extra steps or config to set up openmrs?

**Comment by witash** [comment]
`POST openhim-core:5001/mediator/openmrs/patient` this is not used anymore
for requests to openmrs [this commit](https://github.com/medic/cht-interoperability/pull/115/commits/613bf6b064eafde72ce68e001e3532a60a955aca) gets the configurator set up scripts to add the OpenMRS channel automatically. but not going through any mediator so just like `POST openhim-core:5001/openmrs/patient`

**Comment by andrablaj** [comment]
After discussion with @witash, he mentioned some cases that would need tests, once the current e2e tests in the `openmrs-mediator` branch pass:

- Tests for the edge cases in the sync process, e.g. ensuring no duplicates are created.
- e2e tests for reports.

**Comment by andrablaj** [comment]
I vote to open a new ticket with the new requirement (but I don't feel strongly about it 😄 )
